### ISSUE TITLE
Prioritizing Pods in Scheduling 

### ISSUE THREAD

**Issue by astropuffin** [kubernetes/kubernetes#28928]
Hey all, I’m looking to extend the default scheduler with a capability that I don’t think exists yet. I wanted to ask here just in case I'm duplicating work.

My workload:
A few services/forever pods, but a large number of terminating pods (not yet using the jobs object, but that’s the general idea). These jobs can run anywhere from 5 min to 3 days, and we pretty much always have more “requests” of the cluster than capacity (this is by design). We currently have a fairly hacky way of setting up a priority queue, where certain jobs are submitted to the cluster before others. The details of this hacky system are not that relevant for this discussion.

The need:
It is my task to rewrite this system to be more robust. While my use case in prod will be a more complicated and dynamic, a more or less FIFO queue is sufficient to discuss the scheduler code in question (as long as the priority of a job can be changed later, such as through an annotation on the pod).

Wait, doesn't the scheduler already do this?
Currently the default scheduler effectively does a for loop over all pods, which somewhat randomly selects pods to schedule in order. If it can't schedule a pod, it will move onto the next one. I don't really want to touch the logic of determining which _node_ to schedule the pod on; I just want to modify which _pods_ are selected to be scheduled. For my needs, the scheduler doesn't have to be lightning fast as long as it's reasonable.

Gotchas
Large request pods: It's possible that small pods can get blocked behind large pods in the queue. I have a straightforward way to overcome this, but that adds more complexity to this issue and I'd rather wait until later.

relevant issues:
Multiple schedulers (wip)
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multiple-schedulers.md
Dispatcher feature request
https://github.com/kubernetes/kubernetes/pull/17430

None of the above has been implemented, and frankly I'm ok with just extending my existing scheduler (and running all pods through this mechanism) until multiple schedulers are supported.

relevant issue labels: area/scheduling

relevant code:
I believe all I have to do is implement this function differently:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L66
Is that it or am I missing something? I am also unclear of where that actually is calculated... 

Can anyone give me a few pointers/suggestions and I'll code it up?
Once I code it up, where should I contribute this back to? I'm sure I'm not the only one trying to schedule a ton of jobs on a cluster.

**Comment by therc** [comment]
You also need rescheduling and disruption budgets, for high-priority pods that arrive after low-priority ones, which will is bound to happen at some point. See also https://github.com/kubernetes/kubernetes/pull/22217, https://github.com/kubernetes/kubernetes/issues/22212, https://github.com/kubernetes/kubernetes/issues/23781, https://github.com/kubernetes/kubernetes/issues/12140, etc.
Once that's taken care of, priority-based scheduling can happen (I hope we get it in 1.4).

**Comment by astropuffin** [comment]
Thanks @therc ! Now that I've had a chance to review the linked issues, I have a couple of comments (below). Also bringing in @davidopp to the discussion.

More context on my specific goal: in order to reduce the average cycle time for jobs, I want to implement a [highest response ratio next](https://en.wikipedia.org/wiki/Highest_response_ratio_next) queue. The current system is effectively random in an over capacity cluster. I can totally see others coming up with more complex priority functions. The scope of this issue is not how to calculate the priority, but how to use it given that it can change over relatively short time periods (minutes).

#22217 : Great work on a very complex topic. My use cause is considerably more narrow, and I don't believe needs to depend on the ability to _reschedule_ a pod. Dynamically reshuffling running pods based on a priority is incredibly hard to get right, and isn't what I want to do. (see next).

#22212 , #23781, #12140 : All of my pods are essentially running in the same "QoS band", but some are much more urgent than others. It is incredibly costly for the business if a job is terminated before completion (not only in compute resources, but in time delay as a computation is restarted from scratch), so rescheduling is very much _not_ what I want to do. Even a "low priority" pod should run to completion.

Re: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduling.md#priority
Named strings are not going to work for my use case _at all_. My current hacky solution does just that. It seemed like a great idea at the time,  but it has become a nightmare to manage the increasing number of discrete "levels". As more levels get added their names (like "high", "highest", "default", "admin") because increasingly less meaningful. I need to programmatically reorder the queue into an arbitrary number of (possibly non-unique) priorities. It is _far_ easier to do so with integers.

I can totally see the desire to have named "QoS bands" like "guaranteed" or "interruptible". Priority needs to be much more granular than that though, and I think they should be distinct concepts. I see value in an api object that points to named "priority numbers" for some syntactic sugar, although these should be decidedly optional. You can even have some helper function to "re-annotate" if you want to reorder the priority level names. This seems to be to just be asking for trouble (and unneeded complexity in kubernetes), and is probably best left to the user as it is simple enough to write a one-off script.

I have never worked with Borg/Omega, so I'd love to get more experienced opinions on this.

More Relevant code:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L66
->
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L336
-> getNextPod()
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L416

**Comment by davidopp** [comment]
I chatted with @astropuffin on Slack to unblock him for working on something that meets his requirements, but with the understanding it wouldn't be upstreamed as-is. I'll reply to this issue when I get more time.

**Comment by astropuffin** [comment]
Hey @davidopp,
I have opened a sample implementation PR (https://github.com/kubernetes/kubernetes/pull/31809). The PR is in no way viable for merging, but it does accomplish the task at hand. There are obviously a number of hacks, but I wanted to get your thoughts on it.

Based on this work, in order for the scheduler to support this feature, the scheduler factory would have to be refactored a bit, and the scheduler code would need to reenqueue a pod object if it fails to schedule. I think this would have an enormously detrimental effect on the speed of the scheduler in the general case, but I'm still a bit lost in the complexity of the abstraction in the scheduler so I didn't end up finding a more performant solution.

**Comment by astropuffin** [comment]
I closed out the PR with the sample implementation, because there is a more robust version in flight by the scheduling sig. Leaving this ticket open until the corresponding controllers get merged in.
